File: packages/go-verify/currency_test.go
// SPDX-License-Identifier: Apache-2.0
//
// EP-CURRENCY-v1 Go parity tests. Assert the SAME two-valued verification result
// as packages/verify/currency.test.js: authentic_as_of_commit passes through, and
// currency_at_T is the COMPUTED value offline verification cannot supply. Covers
// all four status outcomes plus every fail-safe branch, matching the JS reason
// strings exactly.
package emiliaverify

import (
	"sort"
	"testing"
	"time"
)

const curNow = "2026-07-05T12:00:00.000Z"

var (
	curActionHash = "sha256:" + repeat("a", 64)
	curOtherHash  = "sha256:" + repeat("b", 64)
)

func repeat(s string, n int) string {
	out := make([]byte, 0, len(s)*n)
	for i := 0; i < n; i++ {
		out = append(out, s...)
	}
	return string(out)
}

func f64(v float64) *float64 { return &v }
func sp(v string) *string    { return &v }

// headAt returns a head observed sec seconds before curNow.
func headAt(sec int, extra func(*FreshHead)) *FreshHead {
	base, _ := time.Parse(time.RFC3339Nano, curNow)
	observed := base.Add(-time.Duration(sec) * time.Second).UTC().Format("2006-01-02T15:04:05.000Z07:00")
	h := &FreshHead{ObservedAt: observed}
	if extra != nil {
		extra(h)
	}
	return h
}

func curReceipt() *CurrencyReceipt { return &CurrencyReceipt{ActionHash: curActionHash} }

func TestCurrencyEnumAndVersion(t *testing.T) {
	got := append([]string{}, CurrencyStatus...)
	sort.Strings(got)
	want := []string{"fresh", "stale", "unknown"}
	for i := range want {
		if got[i] != want[i] {
			t.Fatalf("CurrencyStatus=%v want %v", got, want)
		}
	}
	if CurrencyVersion != "EP-CURRENCY-v1" {
		t.Fatalf("CurrencyVersion=%q", CurrencyVersion)
	}
}

func TestCurrencyNoFreshHeadUnknown(t *testing.T) {
	r := EvaluateCurrency(CurrencyArgs{Receipt: curReceipt(), AuthenticAsOfCommit: true, Now: sp(curNow)})
	if !r.AuthenticAsOfCommit {
		t.Fatal("authentic should pass through true")
	}
	if r.CurrencyAtT.Status != "unknown" {
		t.Fatalf("status=%q want unknown", r.CurrencyAtT.Status)
	}
	if r.CurrencyAtT.Reason != CurrencyReasonOfflineOnlyNoFreshHead {
		t.Fatalf("reason=%q", r.CurrencyAtT.Reason)
	}
	if r.CurrencyAtT.EvaluatedAt == nil || *r.CurrencyAtT.EvaluatedAt != curNow {
		t.Fatalf("evaluated_at=%v want %q", r.CurrencyAtT.EvaluatedAt, curNow)
	}
}

func TestCurrencyFreshWithinWindow(t *testing.T) {
	r := EvaluateCurrency(CurrencyArgs{
		Receipt: curReceipt(), AuthenticAsOfCommit: true, Now: sp(curNow),
		MaxStalenessSeconds: f64(300), FreshHead: headAt(60, nil),
	})
	if r.CurrencyAtT.Status != "fresh" {
		t.Fatalf("status=%q want fresh", r.CurrencyAtT.Status)
	}
	if r.CurrencyAtT.Reason != CurrencyReasonFreshHeadWithinWindow {
		t.Fatalf("reason=%q", r.CurrencyAtT.Reason)
	}
	if r.CurrencyAtT.EvaluatedAt == nil || *r.CurrencyAtT.EvaluatedAt != curNow {
		t.Fatalf("evaluated_at=%v", r.CurrencyAtT.EvaluatedAt)
	}
}

func TestCurrencyFreshIndependentOfAuthenticity(t *testing.T) {
	r := EvaluateCurrency(CurrencyArgs{
		Receipt: curReceipt(), AuthenticAsOfCommit: false, Now: sp(curNow),
		MaxStalenessSeconds: f64(300), FreshHead: headAt(10, nil),
	})
	if r.AuthenticAsOfCommit {
		t.Fatal("authentic should be false")
	}
	if r.CurrencyAtT.Status != "fresh" {
		t.Fatalf("status=%q want fresh", r.CurrencyAtT.Status)
	}
}

func TestCurrencyStaleTooOld(t *testing.T) {
	r := EvaluateCurrency(CurrencyArgs{
		Receipt: curReceipt(), AuthenticAsOfCommit: true, Now: sp(curNow),
		MaxStalenessSeconds: f64(300), FreshHead: headAt(600, nil),
	})
	if r.CurrencyAtT.Status != "stale" || r.CurrencyAtT.Reason != CurrencyReasonFreshHeadStale {
		t.Fatalf("got %+v", r.CurrencyAtT)
	}
}

func TestCurrencyFutureHeadStale(t *testing.T) {
	r := EvaluateCurrency(CurrencyArgs{
		Receipt: curReceipt(), AuthenticAsOfCommit: true, Now: sp(curNow),
		MaxStalenessSeconds: f64(300), FreshHead: headAt(-60, nil),
	})
	if r.CurrencyAtT.Status != "stale" || r.CurrencyAtT.Reason != CurrencyReasonFreshHeadInFuture {
		t.Fatalf("got %+v", r.CurrencyAtT)
	}
}

func TestCurrencyRevokedScalar(t *testing.T) {
	r := EvaluateCurrency(CurrencyArgs{
		Receipt: curReceipt(), AuthenticAsOfCommit: true, Now: sp(curNow),
		MaxStalenessSeconds: f64(300),
		FreshHead:           headAt(5, func(h *FreshHead) { h.Revoked = true }),
	})
	if r.CurrencyAtT.Status != "stale" || r.CurrencyAtT.Reason != CurrencyReasonRevokedByFreshHead {
		t.Fatalf("got %+v", r.CurrencyAtT)
	}
}

func TestCurrencyRevokedByStatusList(t *testing.T) {
	r := EvaluateCurrency(CurrencyArgs{
		Receipt: curReceipt(), AuthenticAsOfCommit: true, Now: sp(curNow),
		MaxStalenessSeconds: f64(300),
		FreshHead:           headAt(5, func(h *FreshHead) { h.RevokedTargetHashes = []string{curActionHash} }),
	})
	if r.CurrencyAtT.Status != "stale" || r.CurrencyAtT.Reason != CurrencyReasonRevokedByFreshHead {
		t.Fatalf("got %+v", r.CurrencyAtT)
	}
}

func TestCurrencyStatusListDifferentTargetStaysFresh(t *testing.T) {
	r := EvaluateCurrency(CurrencyArgs{
		Receipt: curReceipt(), AuthenticAsOfCommit: true, Now: sp(curNow),
		MaxStalenessSeconds: f64(300),
		FreshHead:           headAt(5, func(h *FreshHead) { h.RevokedTargetHashes = []string{curOtherHash} }),
	})
	if r.CurrencyAtT.Status != "fresh" {
		t.Fatalf("status=%q want fresh", r.CurrencyAtT.Status)
	}
}

func TestCurrencyRequiredButAbsent(t *testing.T) {
	r := EvaluateCurrency(CurrencyArgs{
		Receipt: curReceipt(), AuthenticAsOfCommit: true, Now: sp(curNow),
		FreshHeadRequired: true,
	})
	if r.CurrencyAtT.Status != "stale" || r.CurrencyAtT.Reason != CurrencyReasonFreshHeadRequiredButAbsent {
		t.Fatalf("got %+v", r.CurrencyAtT)
	}
}

func TestCurrencyNoPolicyBound(t *testing.T) {
	r := EvaluateCurrency(CurrencyArgs{
		Receipt: curReceipt(), AuthenticAsOfCommit: true, Now: sp(curNow),
		FreshHead: headAt(1, nil), // no MaxStalenessSeconds
	})
	if r.CurrencyAtT.Status != "stale" || r.CurrencyAtT.Reason != CurrencyReasonMaxStalenessInvalid {
		t.Fatalf("got %+v", r.CurrencyAtT)
	}
}

func TestCurrencyNegativeBound(t *testing.T) {
	r := EvaluateCurrency(CurrencyArgs{
		Receipt: curReceipt(), AuthenticAsOfCommit: true, Now: sp(curNow),
		MaxStalenessSeconds: f64(-1), FreshHead: headAt(1, nil),
	})
	if r.CurrencyAtT.Status != "stale" || r.CurrencyAtT.Reason != CurrencyReasonMaxStalenessInvalid {
		t.Fatalf("got %+v", r.CurrencyAtT)
	}
}

func TestCurrencyUnparseableNow(t *testing.T) {
	r := EvaluateCurrency(CurrencyArgs{
		Receipt: curReceipt(), AuthenticAsOfCommit: true, Now: sp("not-a-time"),
		MaxStalenessSeconds: f64(300), FreshHead: headAt(1, nil),
	})
	if r.CurrencyAtT.Status != "unknown" || r.CurrencyAtT.Reason != CurrencyReasonNowInvalid {
		t.Fatalf("got %+v", r.CurrencyAtT)
	}
	if r.CurrencyAtT.EvaluatedAt != nil {
		t.Fatalf("evaluated_at should be nil, got %v", *r.CurrencyAtT.EvaluatedAt)
	}
}

func TestCurrencyMalformedHead(t *testing.T) {
	// head with no observation instant.
	r := EvaluateCurrency(CurrencyArgs{
		Receipt: curReceipt(), AuthenticAsOfCommit: true, Now: sp(curNow),
		MaxStalenessSeconds: f64(300), FreshHead: &FreshHead{},
	})
	if r.CurrencyAtT.Status != "unknown" || r.CurrencyAtT.Reason != CurrencyReasonFreshHeadMalformed {
		t.Fatalf("got %+v", r.CurrencyAtT)
	}
}

func TestCurrencyAuthenticPassThroughAndFailSafe(t *testing.T) {
	// authentic omitted (zero value false).
	r := EvaluateCurrency(CurrencyArgs{Receipt: curReceipt(), Now: sp(curNow)})
	if r.AuthenticAsOfCommit {
		t.Fatal("omitted authentic should be false")
	}
}

func TestCurrencyEmptyArgsFailSafeDefault(t *testing.T) {
	r := EvaluateCurrency(CurrencyArgs{})
	if r.AuthenticAsOfCommit {
		t.Fatal("authentic default false")
	}
	if r.CurrencyAtT.Status != "unknown" || r.CurrencyAtT.Reason != CurrencyReasonOfflineOnlyNoFreshHead {
		t.Fatalf("got %+v", r.CurrencyAtT)
	}
}

File: packages/verify/currency.test.ts
// SPDX-License-Identifier: Apache-2.0
//
// EP-CURRENCY-v1 test. Asserts the two-valued verification result:
// authentic_as_of_commit is passed through, and currency_at_T is the COMPUTED
// value offline verification cannot supply. Covers all four status outcomes:
//   - 'unknown' — no fresh head (the honest, fail-safe offline default)
//   - 'fresh'   — a recent, non-revoking head within the policy staleness bound
//   - 'stale'   — head older than the policy bound
//   - 'stale'   (revoked) — head shows this receipt revoked
// Plus the fail-safe branches: required-but-absent head, missing policy bound,
// unparseable now, malformed head, and offline-false pass-through.
import { test } from 'node:test';
import assert from 'node:assert';
import { evaluateCurrency, CURRENCY_STATUS, CURRENCY_REASON, CURRENCY_VERSION } from './currency.js';

const NOW = '2026-07-05T12:00:00.000Z';
const ACTION_HASH = 'sha256:' + 'a'.repeat(64);
const OTHER_HASH = 'sha256:' + 'b'.repeat(64);
const receipt = { action_hash: ACTION_HASH };

// A head observed `sec` seconds before NOW.
function headAt(sec, extra = {}) {
  const observed_at = new Date(Date.parse(NOW) - sec * 1000).toISOString();
  return { observed_at, ...extra };
}

test('enum and version are the honest two-valued shape', () => {
  assert.deepStrictEqual([...CURRENCY_STATUS].sort(), ['fresh', 'stale', 'unknown']);
  assert.strictEqual(CURRENCY_VERSION, 'EP-CURRENCY-v1');
});

// ── status: 'unknown' — the fail-safe offline default ────────────────────────
test("no freshHead => 'unknown' (offline cannot prove currency, never 'fresh')", () => {
  const r = evaluateCurrency({ receipt, authentic_as_of_commit: true, now: NOW });
  assert.strictEqual(r.authentic_as_of_commit, true);
  assert.strictEqual(r.currency_at_T.status, 'unknown');
  assert.strictEqual(r.currency_at_T.reason, CURRENCY_REASON.offline_only_no_fresh_head);
  assert.strictEqual(r.currency_at_T.evaluated_at, NOW);
  // The core honesty invariant: an offline-only check must NEVER report 'fresh'.
  assert.notStrictEqual(r.currency_at_T.status, 'fresh');
});

test("null freshHead is the same fail-safe 'unknown' as absent", () => {
  const r = evaluateCurrency({ receipt, authentic_as_of_commit: true, now: NOW, freshHead: null });
  assert.strictEqual(r.currency_at_T.status, 'unknown');
  assert.strictEqual(r.currency_at_T.reason, CURRENCY_REASON.offline_only_no_fresh_head);
});

// ── status: 'fresh' — recent, non-revoking head within policy bound ──────────
test("recent non-revoking head within maxStalenessSeconds => 'fresh'", () => {
  const r = evaluateCurrency({
    receipt,
    authentic_as_of_commit: true,
    now: NOW,
    maxStalenessSeconds: 300,
    freshHead: headAt(60), // 60s old, bound is 300s
  });
  assert.strictEqual(r.currency_at_T.status, 'fresh');
  assert.strictEqual(r.currency_at_T.reason, CURRENCY_REASON.fresh_head_within_window);
  assert.strictEqual(r.currency_at_T.evaluated_at, NOW);
});

test("'fresh' does not depend on authenticity — currency is a separate axis", () => {
  // authentic_as_of_commit passes through independently of currency.
  const r = evaluateCurrency({
    receipt,
    authentic_as_of_commit: false,
    now: NOW,
    maxStalenessSeconds: 300,
    freshHead: headAt(10),
  });
  assert.strictEqual(r.authentic_as_of_commit, false);
  assert.strictEqual(r.currency_at_T.status, 'fresh');
});

// ── status: 'stale' — head older than the policy bound ───────────────────────
test("head older than maxStalenessSeconds => 'stale'", () => {
  const r = evaluateCurrency({
    receipt,
    authentic_as_of_commit: true,
    now: NOW,
    maxStalenessSeconds: 300,
    freshHead: headAt(600), // 600s old, bound is 300s
  });
  assert.strictEqual(r.currency_at_T.status, 'stale');
  assert.strictEqual(r.currency_at_T.reason, CURRENCY_REASON.fresh_head_stale);
});

test("future-dated status head => 'stale', never 'fresh'", () => {
  const r = evaluateCurrency({
    receipt,
    authentic_as_of_commit: true,
    now: NOW,
    maxStalenessSeconds: 300,
    freshHead: headAt(-60),
  });
  assert.strictEqual(r.currency_at_T.status, 'stale');
  assert.strictEqual(r.currency_at_T.reason, CURRENCY_REASON.fresh_head_in_future);
});

// ── status: 'stale' — head shows revocation ──────────────────────────────────
test("head with scalar revoked:true => 'stale' (revoked), even if recent", () => {
  const r = evaluateCurrency({
    receipt,
    authentic_as_of_commit: true,
    now: NOW,
    maxStalenessSeconds: 300,
    freshHead: headAt(5, { revoked: true }),
  });
  assert.strictEqual(r.currency_at_T.status, 'stale');
  assert.strictEqual(r.currency_at_T.reason, CURRENCY_REASON.revoked_by_fresh_head);
});

test("head status-list revoking this receipt's action_hash => 'stale' (revoked)", () => {
  const r = evaluateCurrency({
    receipt,
    authentic_as_of_commit: true,
    now: NOW,
    maxStalenessSeconds: 300,
    freshHead: headAt(5, { revoked_target_hashes: [ACTION_HASH] }),
  });
  assert.strictEqual(r.currency_at_T.status, 'stale');
  assert.strictEqual(r.currency_at_T.reason, CURRENCY_REASON.revoked_by_fresh_head);
});

test("status-list that revokes a DIFFERENT target does not revoke this receipt", () => {
  const r = evaluateCurrency({
    receipt,
    authentic_as_of_commit: true,
    now: NOW,
    maxStalenessSeconds: 300,
    freshHead: headAt(5, { revoked_target_hashes: [OTHER_HASH] }),
  });
  assert.strictEqual(r.currency_at_T.status, 'fresh');
});

// ── fail-safe branches ───────────────────────────────────────────────────────
test("policy requires a fresh head but none supplied => 'stale' (required-but-absent)", () => {
  const r = evaluateCurrency({
    receipt,
    authentic_as_of_commit: true,
    now: NOW,
    freshHeadRequired: true,
  });
  assert.strictEqual(r.currency_at_T.status, 'stale');
  assert.strictEqual(r.currency_at_T.reason, CURRENCY_REASON.fresh_head_required_but_absent);
});

test("freshHead supplied but no policy bound => 'stale' (cannot certify freshness)", () => {
  const r = evaluateCurrency({
    receipt,
    authentic_as_of_commit: true,
    now: NOW,
    freshHead: headAt(1), // recent, but no maxStalenessSeconds to measure against
  });
  assert.strictEqual(r.currency_at_T.status, 'stale');
  assert.strictEqual(r.currency_at_T.reason, CURRENCY_REASON.max_staleness_invalid);
});

test("negative maxStalenessSeconds is not a valid bound => 'stale'", () => {
  const r = evaluateCurrency({
    receipt,
    authentic_as_of_commit: true,
    now: NOW,
    maxStalenessSeconds: -1,
    freshHead: headAt(1),
  });
  assert.strictEqual(r.currency_at_T.status, 'stale');
  assert.strictEqual(r.currency_at_T.reason, CURRENCY_REASON.max_staleness_invalid);
});

test("non-finite maxStalenessSeconds is not a valid bound => 'stale'", () => {
  for (const bound of [Number.NaN, Number.POSITIVE_INFINITY]) {
    const r = evaluateCurrency({
      receipt,
      authentic_as_of_commit: true,
      now: NOW,
      maxStalenessSeconds: bound,
      freshHead: headAt(1),
    });
    assert.strictEqual(r.currency_at_T.status, 'stale');
    assert.strictEqual(r.currency_at_T.reason, CURRENCY_REASON.max_staleness_invalid);
  }
});

test("unparseable now => 'unknown' (won't measure age against a bad clock)", () => {
  const r = evaluateCurrency({
    receipt,
    authentic_as_of_commit: true,
    now: 'not-a-time',
    maxStalenessSeconds: 300,
    freshHead: headAt(1),
  });
  assert.strictEqual(r.currency_at_T.status, 'unknown');
  assert.strictEqual(r.currency_at_T.reason, CURRENCY_REASON.now_invalid);
  assert.strictEqual(r.currency_at_T.evaluated_at, null);
});

test("malformed head (no observation instant) => 'unknown', never 'fresh'", () => {
  const r = evaluateCurrency({
    receipt,
    authentic_as_of_commit: true,
    now: NOW,
    maxStalenessSeconds: 300,
    freshHead: { revoked: false }, // no observed_at / issued_at
  });
  assert.strictEqual(r.currency_at_T.status, 'unknown');
  assert.strictEqual(r.currency_at_T.reason, CURRENCY_REASON.fresh_head_malformed);
});

test('authentic_as_of_commit passes through and fail-safes non-true to false', () => {
  const r1 = evaluateCurrency({ receipt, authentic_as_of_commit: 'yes', now: NOW });
  assert.strictEqual(r1.authentic_as_of_commit, false);
  const r2 = evaluateCurrency({ receipt, now: NOW }); // omitted
  assert.strictEqual(r2.authentic_as_of_commit, false);
});

test('empty args object does not throw and yields the fail-safe unknown default', () => {
  const r = evaluateCurrency();
  assert.strictEqual(r.authentic_as_of_commit, false);
  assert.strictEqual(r.currency_at_T.status, 'unknown');
  assert.strictEqual(r.currency_at_T.reason, CURRENCY_REASON.offline_only_no_fresh_head);
});
